### PR TITLE
Extend s3 support to datasets def and data

### PR DESCRIPTION
**Pull request by afrittoli**
Allow for dataset definition and normalized data to be store in s3
as well. This normalized data can act as input to a cloud training
system such as kubeflow, ffdl or ciml own.

### CODE REVIEW

File: ciml/gather_results.py [hunk 1]
```diff
@@ -532,23 +534,48 @@ def get_experiment_folder(dataset, name, data_path=None):
...
+    dataset_file = get_data_path(data_path=data_path, s3=s3)
+    dataset_file.append(dataset)
+    use_s3 = (dataset_file[0] == 's3:')
+
+    if use_s3:
+        # Download as a file first. Not that efficient, but we avoid having
+        # to keep the whole dataset in memory.
+        object_key = os.sep.join(dataset_file[3:] + [name + '.npz'])
+        _, target_filename = tempfile.mkstemp(prefix=None, suffix='.npz')
+        s3.download_file(Bucket=dataset_file[2], Key=object_key,
```
Comment (mtreinish) [inline ciml/gather_results.py:554]: What if s3 is None here? Don't we need to initialize a client?

File: ciml/gather_results.py [hunk 2]
```diff
@@ -532,23 +534,48 @@ def get_experiment_folder(dataset, name, data_path=None):
...
     else:
-        dataset_file = [data_path]
-    dataset_file.extend(['data', dataset, name])
-    np.savez_compressed(os.sep.join(dataset_file), **kwargs)
+        target_filename = os.sep.join(dataset_file)
+    np.savez_compressed(target_filename, **kwargs)
+
+    if use_s3:
+        object_key = os.sep.join(dataset_file[3:] + [name + '.npz'])
+        s3.upload_file(Filename=target_filename, Bucket=dataset_file[2],
```
Comment (mtreinish) [inline ciml/gather_results.py:583]: Same issue as above, what if s3 is None and the path starts with 's3:'

**Review by mtreinish — COMMENTED** [review]
LGTM, but I spotted one potential issue. Since we've decoupled use_s3 and the actual client s3 definition we could potentially trying call an s3 client without getting one passed in

File: ciml/gather_results.py (hunk 1 above)
Comment (afrittoli) [reply ciml/gather_results.py:554]: There are two ways we can solve this, either fail is s3 is not passed and data_path is s3://, or else we can get an s3 client on the fly, only we don't have the s3 parameters, so we can get a vanilla one which may not be the right thing to do. So probably we should fail.

Comment (afrittoli) [reply ciml/gather_results.py:554]: Updated the code, throwing ValueError now.

File: ciml/gather_results.py (hunk 2 above)
Comment (afrittoli) [reply ciml/gather_results.py:583]: Updated the code, throwing ValueError now.

**Review by mtreinish — APPROVED** [review]
ok, LGTM now

**Pull request merged**